File: internal/slackagent/canvas_publisher.go
package slackagent

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"slices"
	"strings"

	"github.com/google/uuid"
)

const defaultCanvasOutDir = "./runtime/canvases"

type CanvasArtifact struct {
	ID      string                 `json:"id,omitempty"`
	Title   string                 `json:"title,omitempty"`
	MeetURL string                 `json:"meet_url,omitempty"`
	Files   *CanvasArtifactFiles   `json:"files,omitempty"`
	Summary *CanvasArtifactSummary `json:"summary,omitempty"`
}

type CanvasArtifactFiles struct {
	Transcript string `json:"transcript,omitempty"`
	Audio      string `json:"audio,omitempty"`
	Summary    string `json:"summary,omitempty"`
}

type CanvasArtifactSummary struct {
	Highlights       []string `json:"highlights,omitempty"`
	Decisions        []string `json:"decisions,omitempty"`
	ActionItems      []string `json:"action_items,omitempty"`
	Attendees        []string `json:"attendees,omitempty"`
	DurationMinutes  int      `json:"duration_minutes,omitempty"`
	DurationMinutes2 int      `json:"durationMinutes,omitempty"`
}

type CanvasPublisherConfig struct {
	Provider   string
	OutDir     string
	Channel    string
	ThreadTS   string
	BotToken   string
	APIBaseURL string
	Client     *http.Client
	Poster     *Poster
}

type CanvasPublishInput struct {
	Artifact         CanvasArtifact
	ArtifactID       string
	ID               string
	Title            string
	SummaryMarkdown  string
	SummaryPath      string
	NotificationText string
	Destination      string
	Channel          string
	ThreadTS         string
	DedupKey         string
	CanvasID         string
	Operation        string
	SectionID        string
	ForceSlackCanvas bool
}

type PublishedCanvasManifest struct {
	ID           string                `json:"id"`
	Provider     string                `json:"provider"`
	Surface      string                `json:"surface"`
	ArtifactID   string                `json:"artifact_id"`
	MarkdownPath string                `json:"markdown_path"`
	ManifestPath string                `json:"manifest_path"`
	CreatedAt    string                `json:"created_at"`
	OK           bool                  `json:"ok"`
	Destination  string                `json:"destination"`
	Slack        *PostMessageResult    `json:"slack,omitempty"`
	Canvas       *SlackCanvasAPIResult `json:"canvas,omitempty"`
}

type CanvasPublisher struct {
	provider   string
	outDir     string
	channel    string
	threadTS   string
	botToken   string
	apiBaseURL string
	client     *http.Client
	poster     *Poster
}

func NewCanvasPublisher(config CanvasPublisherConfig) (*CanvasPublisher, error) {
	outDir := strings.TrimSpace(config.OutDir)
	if outDir == "" {
		outDir = defaultCanvasOutDir
	}
	if err := os.MkdirAll(outDir, 0o755); err != nil {
		return nil, fmt.Errorf("create canvas out dir: %w", err)
	}

	provider := normalizeCanvasProvider(config.Provider)
	poster := config.Poster
	if poster == nil {
		poster = NewPoster(PosterConfig{
			BotToken: config.BotToken,
			Mock:     strings.TrimSpace(config.BotToken) == "",
			Client:   config.Client,
		})
	}

	return &CanvasPublisher{
		provider:   provider,
		outDir:     outDir,
		channel:    strings.TrimSpace(config.Channel),
		threadTS:   strings.TrimSpace(config.ThreadTS),
		botToken:   strings.TrimSpace(config.BotToken),
		apiBaseURL: strings.TrimSpace(config.APIBaseURL),
		client:     config.Client,
		poster:     poster,
	}, nil
}

func (p *CanvasPublisher) Publish(ctx context.Context, input CanvasPublishInput) (PublishedCanvasManifest, error) {
	artifactID := firstNonEmpty(input.ArtifactID, input.Artifact.ID, "meeting-"+uuid.NewString()[:8])
	id := firstNonEmpty(input.ID, "canvas-"+sanitizeCanvasID(artifactID)+"-"+uuid.NewString()[:8])
	createdAt := nowRFC3339()
	markdown, err := renderCanvasMarkdown(input)
	if err != nil {
		return PublishedCanvasManifest{}, err
	}

	markdownPath := filepath.Join(p.outDir, id+".md")
	manifestPath := filepath.Join(p.outDir, id+".json")
	if err := os.WriteFile(markdownPath, []byte(markdown), 0o644); err != nil {
		return PublishedCanvasManifest{}, fmt.Errorf("write canvas markdown: %w", err)
	}

	result := PublishedCanvasManifest{
		ID:           id,
		Provider:     p.provider,
		Surface:      p.provider,
		ArtifactID:   artifactID,
		MarkdownPath: markdownPath,
		ManifestPath: manifestPath,
		CreatedAt:    createdAt,
		OK:           true,
		Destination:  strings.TrimSpace(input.Destination),
	}

	switch {
	case p.provider == "slack-canvas" || input.ForceSlackCanvas:
		canvasResult := p.publishSlackCanvas(ctx, markdown, input)
		result.OK = canvasResult.OK
		result.Surface = "slack-canvas"
		result.Canvas = &canvasResult
		postChannel := firstNonEmpty(input.Channel, p.channel)
		if canvasResult.OK && postChannel != "" && strings.TrimSpace(input.NotificationText) != "" {
			postResult := p.poster.PostMessage(ctx, PostMessageInput{
				Channel:  postChannel,
				ThreadTS: firstNonEmpty(input.ThreadTS, p.threadTS),
				Text:     strings.ReplaceAll(input.NotificationText, "{{canvas_link}}", slackCanvasMarkdownLink(canvasResult)),
				DedupKey: firstNonEmpty(input.DedupKey, defaultCanvasDedupKey(artifactID, firstNonEmpty(input.Channel, p.channel), firstNonEmpty(input.ThreadTS, p.threadTS))),
			})
			result.Slack = &postResult
			result.OK = result.OK && postResult.OK
		}
	case p.provider == "slack-thread" || strings.TrimSpace(input.Channel) != "":
		postResult := p.poster.PostMessage(ctx, PostMessageInput{
			Channel:  firstNonEmpty(input.Channel, p.channel),
			ThreadTS: firstNonEmpty(input.ThreadTS, p.threadTS),
			Text:     truncateSlackText(markdown),
			DedupKey: firstNonEmpty(input.DedupKey, defaultCanvasDedupKey(artifactID, firstNonEmpty(input.Channel, p.channel), firstNonEmpty(input.ThreadTS, p.threadTS))),
		})
		result.OK = postResult.OK
		if postResult.Mock {
			result.Surface = "mock-slack-thread"
		} else {
			result.Surface = "slack-thread"
		}
		result.Slack = &postResult
	}

	if err := writeJSONFile(manifestPath, result); err != nil {
		return PublishedCanvasManifest{}, err
	}
	return result, nil
}

func (p *CanvasPublisher) ListPublished() ([]PublishedCanvasManifest, error) {
	entries, err := os.ReadDir(p.outDir)
	if err != nil {
		return nil, fmt.Errorf("read canvas out dir: %w", err)
	}

	manifests := make([]PublishedCanvasManifest, 0)
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".json") {
			continue
		}
		raw, err := os.ReadFile(filepath.Join(p.outDir, entry.Name()))
		if err != nil {
			return nil, fmt.Errorf("read canvas manifest: %w", err)
		}
		var manifest PublishedCanvasManifest
		if err := json.Unmarshal(raw, &manifest); err != nil {
			return nil, fmt.Errorf("parse canvas manifest: %w", err)
		}
		manifests = append(manifests, manifest)
	}

	slices.SortFunc(manifests, func(a, b PublishedCanvasManifest) int {
		return strings.Compare(a.CreatedAt, b.CreatedAt)
	})
	return manifests, nil
}

func (p *CanvasPublisher) publishSlackCanvas(ctx context.Context, markdown string, input CanvasPublishInput) SlackCanvasAPIResult {
	if strings.TrimSpace(input.CanvasID) != "" {
		return EditSlackCanvas(
			ctx,
			p.client,
			p.botToken,
			p.apiBaseURL,
			input.CanvasID,
			markdown,
			input.Operation,
			input.SectionID,
		)
	}
	channel := firstNonEmpty(input.Channel, p.channel)
	if input.ForceSlackCanvas {
		channel = ""
	}
	result := CreateSlackCanvas(
		ctx,
		p.client,
		p.botToken,
		p.apiBaseURL,
		firstNonEmpty(input.Title, input.Artifact.Title, "Meeting summary"),
		markdown,
		channel,
	)
	if result.OK && strings.TrimSpace(result.CanvasID) != "" {
		auth := SlackAuthTest(ctx, p.client, p.botToken, p.apiBaseURL)
		result.TeamID = auth.TeamID
		result.Permalink = slackCanvasPermalink(auth.TeamID, result.CanvasID)
	}
	return result
}

func normalizeCanvasProvider(provider string) string {
	normalized := strings.NewReplacer("_", "-", " ", "-").Replace(strings.ToLower(strings.TrimSpace(provider)))
	if normalized == "" {
		return "file"
	}
	return normalized
}

File: packages/core/src/slack/canvas-publisher.ts
import crypto from "node:crypto";
import { existsSync, mkdirSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import { join, resolve } from "node:path";
import { getRuntimeConfig } from "../env.js";
import { createSlackPoster } from "./slack-poster.js";

export interface CanvasArtifact {
  id?: string;
  title?: string;
  meetUrl?: string;
  files?: { transcript?: string; audio?: string; [key: string]: unknown };
  summary?: {
    highlights?: unknown[];
    decisions?: unknown[];
    actionItems?: unknown[];
    [key: string]: unknown;
  };
  [key: string]: unknown;
}

type FetchLike = typeof fetch;

function normalizeProvider(provider: unknown): string {
  return String(provider || "file")
    .trim()
    .toLowerCase()
    .replaceAll("_", "-");
}

function safeJsonParse<T = unknown>(raw: string, fallback: T | null = null): T | null {
  try {
    return JSON.parse(raw) as T;
  } catch {
    return fallback;
  }
}

function sanitizeId(value: unknown, fallback: string = "canvas"): string {
  const safe = String(value || "")
    .toLowerCase()
    .replace(/[^a-z0-9._-]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 90);
  return safe || fallback;
}

function readTextIfExists(path: string): string {
  if (!path || !existsSync(path)) return "";
  return readFileSync(path, "utf8");
}

export interface CallSlackApiArgs {
  botToken?: string;
  method: string;
  payload: Record<string, unknown>;
  fetchImpl?: FetchLike;
}

export interface CallSlackApiResult {
  ok: boolean;
  status?: number;
  method: string;
  body?: { ok?: boolean; error?: string; canvas_id?: string; canvas?: { id?: string }; [key: string]: unknown };
  error?: string;
  detail?: string;
}

export async function callSlackApi({
  botToken,
  method,
  payload,
  fetchImpl = fetch,
}: CallSlackApiArgs): Promise<CallSlackApiResult> {
  if (!botToken) return { ok: false, error: "missing_slack_bot_token", method };
  try {
    const response = await fetchImpl(`https://slack.com/api/${method}`, {
      method: "POST",
      headers: {
        authorization: `Bearer ${botToken}`,
        "content-type": "application/json; charset=utf-8",
      },
      body: JSON.stringify(payload),
    });
    const body = (await response.json().catch(() => ({}))) as CallSlackApiResult["body"];
    return {
      ok: response.ok && body?.ok === true,
      status: response.status,
      method,
      body,
      error: body?.ok === true ? "" : body?.error || `http_${response.status}`,
    };
  } catch (error) {
    const err = error as { message?: string };
    return {
      ok: false,
      status: 0,
      method,
      error: "request_failed",
      detail: String(err?.message || error),
    };
  }
}

async function createSlackCanvas({
  botToken,
  title,
  markdown,
  channel,
  fetchImpl,
}: {
  botToken?: string;
  title?: string;
  markdown: string;
  channel?: string;
  fetchImpl?: FetchLike;
}) {
  const payload: Record<string, unknown> = {
    title: title || "Meeting summary",
    document_content: {
      type: "markdown",
      markdown,
    },
  };
  if (channel) payload.channel_id = channel;
  const result = await callSlackApi({
    botToken,
    method: "canvases.create",
    payload,
    fetchImpl,
  });
  return {
    ...result,
    canvasId: result.body?.canvas_id || result.body?.canvas?.id || "",
    channel,
  };
}

async function editSlackCanvas({
  botToken,
  canvasId,
  markdown,
  operation = "insert_at_end",
  sectionId = "",
  fetchImpl,
}: {
  botToken?: string;
  canvasId?: string;
  markdown: string;
  operation?: string;
  sectionId?: string;
  fetchImpl?: FetchLike;
}) {
  if (!canvasId) return { ok: false, error: "missing_canvas_id", method: "canvases.edit" };
  const change: Record<string, unknown> = {
    operation,
    document_content: {
      type: "markdown",
      markdown,
    },
  };
  if (sectionId) change.section_id = sectionId;
  return await callSlackApi({
    botToken,
    method: "canvases.edit",
    payload: {
      canvas_id: canvasId,
      changes: [change],
    },
    fetchImpl,
  });
}

function renderPublishMarkdown({
  artifact = {},
  summaryMarkdown = "",
  title = "",
}: {
  artifact?: CanvasArtifact;
  summaryMarkdown?: string;
  title?: string;
}): string {
  if (summaryMarkdown.trim()) return summaryMarkdown;
  const lines = [
    `# ${title || artifact.title || "Meeting summary"}`,
    "",
    `- Meeting: ${artifact.meetUrl || "unknown"}`,
    `- Artifact: ${artifact.id || "unknown"}`,
  ];
  if (artifact.files?.transcript) lines.push(`- Transcript: ${artifact.files.transcript}`);
  if (artifact.files?.audio) lines.push(`- Audio: ${artifact.files.audio}`);
  lines.push("", "## Highlights", "");
  for (const item of artifact.summary?.highlights || []) lines.push(`- ${item}`);
  if (!(artifact.summary?.highlights || []).length) lines.push("- No highlights captured yet.");
  lines.push("", "## Decisions", "");
  for (const item of artifact.summary?.decisions || []) lines.push(`- ${item}`);
  if (!(artifact.summary?.decisions || []).length) lines.push("- No explicit decisions captured.");
  lines.push("", "## Action Items", "");
  for (const item of artifact.summary?.actionItems || []) lines.push(`- ${item}`);
  if (!(artifact.summary?.actionItems || []).length)
    lines.push("- No explicit action items captured.");
  return `${lines.join("\n")}\n`;
}

export interface CreateCanvasPublisherOptions {
  env?: NodeJS.ProcessEnv;
  provider?: string;
  outDir?: string;
  botToken?: string;
  fetchImpl?: FetchLike;
  poster?: ReturnType<typeof createSlackPoster>;
}

export interface CanvasPublishInput {
  artifact?: CanvasArtifact & { files?: { summary?: string } };
  artifactId?: string;
  id?: string;
  title?: string;
  summaryMarkdown?: string;
  summaryPath?: string;
  destination?: string;
  channel?: string;
  threadTs?: string;
  thread_ts?: string;
  canvasId?: string;
  operation?: string;
  sectionId?: string;
  section_id?: string;
  dedupKey?: string;
  [key: string]: unknown;
}

interface PublishedCanvasManifest {
  id: string;
  provider: string;
  surface: string;
  artifactId: string;
  markdownPath: string;
  manifestPath: string;
  createdAt: string;
  ok: boolean;
  destination: string;
  slack: unknown;
  canvas: unknown;
  [key: string]: unknown;
}

export function createCanvasPublisher(options: CreateCanvasPublisherOptions = {}) {
  const config = getRuntimeConfig(options.env);
  const provider = normalizeProvider(options.provider || config.canvasPublisher);
  const outDir = resolve(options.outDir || config.canvasDir);
  const botToken = options.botToken || config.slackBotToken;
  const fetchImpl = options.fetchImpl || fetch;
  const poster =
    options.poster ||
    createSlackPoster({
      botToken,
      mock: config.slackPosterMock || !botToken,
      fetchImpl,
    });
  mkdirSync(outDir, { recursive: true });

  function listPublished(): PublishedCanvasManifest[] {
    if (!existsSync(outDir)) return [];
    return readdirSync(outDir)
      .filter((name) => name.endsWith(".json"))
      .map(
        (name) =>
          safeJsonParse<PublishedCanvasManifest>(readFileSync(join(outDir, name), "utf8")) as
            | PublishedCanvasManifest
            | null,
      )
      .filter((entry): entry is PublishedCanvasManifest => Boolean(entry))
      .sort((a, b) => String(a.createdAt).localeCompare(String(b.createdAt)));
  }

  async function publish(input: CanvasPublishInput = {}): Promise<PublishedCanvasManifest> {
    const artifact = input.artifact || {};
    const artifactId =
      input.artifactId || artifact.id || `meeting-${crypto.randomUUID().slice(0, 8)}`;
    const id = input.id || `canvas-${sanitizeId(artifactId)}-${crypto.randomUUID().slice(0, 8)}`;
    const createdAt = new Date().toISOString();
    const summaryMarkdown =
      input.summaryMarkdown ||
      readTextIfExists(input.summaryPath || artifact.files?.summary || "");
    const markdown = renderPublishMarkdown({
      artifact,
      summaryMarkdown,
      title: input.title,
    });
    const markdownPath = join(outDir, `${id}.md`);
    const manifestPath = join(outDir, `${id}.json`);
    writeFileSync(markdownPath, markdown);

    const result: PublishedCanvasManifest = {
      ok: true,
      id,
      provider,
      surface: provider,
      artifactId,
      markdownPath,
      manifestPath,
      createdAt,
      destination: input.destination || "",
      slack: null,
      canvas: null,
    };

    if (provider === "slack-canvas") {
      const canvas = input.canvasId
        ? await editSlackCanvas({
            botToken,
            canvasId: input.canvasId,
            markdown,
            operation: input.operation,
            sectionId: input.sectionId || input.section_id || "",
            fetchImpl,
          })
        : await createSlackCanvas({
            botToken,
            title: input.title || artifact.title || "Meeting summary",
            markdown,
            channel: input.channel || config.canvasSlackChannel,
            fetchImpl,
          });
      result.ok = canvas.ok;
      result.canvas = canvas;
      result.surface = "slack-canvas";
    } else if (provider === "slack-thread" || provider === "slack" || input.channel) {
      const post = await poster.postMessage({
        channel: input.channel || config.canvasSlackChannel,
        threadTs: input.threadTs || input.thread_ts || config.canvasSlackThreadTs,
        text: markdown.slice(0, 3500),
        dedupKey:
          input.dedupKey ||
          `post-meeting:${artifactId}:${input.channel || config.canvasSlackChannel}:${input.threadTs || input.thread_ts || config.canvasSlackThreadTs}`,
      });
      result.ok = post.ok;
      result.slack = post;
      result.surface = post.mock ? "mock-slack-thread" : "slack-thread";
    }

    writeFileSync(manifestPath, `${JSON.stringify(result, null, 2)}\n`);
    return result;
  }

  return {
    provider,
    outDir,
    publish,
    listPublished,
  };
}
